File: applications/event-account-gemfire-sink/src/main/java/showcase/event/stream/rabbitmq/account/RabbitStreamConfig.java
/*
 *
 *  * Copyright 2023 VMware, Inc.
 *  * SPDX-License-Identifier: GPL-3.0
 *
 */

package showcase.event.stream.rabbitmq.account;

import com.rabbitmq.stream.Consumer;
import com.rabbitmq.stream.Environment;
import com.rabbitmq.stream.OffsetSpecification;
import com.vmware.tanzu.data.services.rabbitmq.streaming.account.domain.Account;
import lombok.extern.slf4j.Slf4j;
import nyla.solutions.core.patterns.conversion.Converter;
import org.springframework.amqp.core.Queue;
import org.springframework.amqp.core.QueueBuilder;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Profile;

import java.util.Arrays;

import static java.lang.String.valueOf;

@Configuration
@Slf4j
@Profile("stream")
public class RabbitStreamConfig {

    private static final String FILTER_PROP_NM = "stateProvince";

    @Value("${spring.application.name}")
    private String applicationName;

    @Value("${spring.rabbitmq.username:guest}")
    private String username = "guest";

    @Value("${spring.rabbitmq.password:guest}")
    private String password = "guest";

    @Value("${spring.rabbitmq.host:127.0.0.1}")
    private String hostname = "localhost";

    @Value("${rabbitmq.streaming.offset:last}")
    private String offset;

    @Value("${rabbitmq.streaming.singleActiveConsumer:false}")
    private boolean singleActiveConsumer;

    @Value("${spring.cloud.stream.bindings.input.destination:event-log-sink}")
    private String streamName;


    @Value("${rabbitmq.streaming.filter.values:}")
    private String[] filterValues;

    @Bean
    Queue stream(Environment environment) {
        log.info("Creating stream: {}",streamName);

        environment.streamCreator().name(streamName)
                .create();

        return QueueBuilder.durable(streamName)
                .stream()
                .build();
    }

    @Bean
    Consumer consumer(Environment environment,
                      java.util.function.Consumer<Account> consumerFunction, Converter<byte[],Account > converter){

        log.info("stream: {}, offset: {}, filterValues: {}, singleActiveConsumer: {}",
                streamName,offset,filterValues,singleActiveConsumer);

        var builder = environment.consumerBuilder()
                .stream(streamName);

        if(filterValues != null && filterValues.length > 0)
                builder = builder.filter().values(filterValues)
                        .postFilter(msg ->
                                Arrays.asList(filterValues)
                                        .contains(valueOf(msg.getApplicationProperties().get(FILTER_PROP_NM))))
                        .builder();

        var rabbitOffset = offset(environment);

        if(singleActiveConsumer)
            builder = builder.name(applicationName)
                    .singleActiveConsumer();

        if(OffsetSpecification.last().equals(rabbitOffset))
            builder = builder.name(applicationName);
        else if(OffsetSpecification.first().equals(rabbitOffset))
        {
            log.info("Replay all messages");
            builder.subscriptionListener(
                    subscriptionContext -> subscriptionContext
                            .offsetSpecification(OffsetSpecification.first()));
        }
        return builder.offset(rabbitOffset)
                .messageHandler((context, message) -> {
                    consumerFunction.accept(converter.convert(message.getBodyAsBinary()));
                })
                .build();
    }

    OffsetSpecification offset(Environment environment){
        return switch (offset)
                    {
                        case "first" -> OffsetSpecification.first();
                        case "next" -> OffsetSpecification.next();
                        default ->  OffsetSpecification.last();
                };
    }
}

File: applications/stream-account-generator-source-perf/src/main/kotlin/com/vmware/tanzu/data/services/rabbitmq/streaming/account/RabbitStreamConfig.kt
package com.vmware.tanzu.data.services.rabbitmq.streaming.account

import com.rabbitmq.stream.ConsumerBuilder
import com.rabbitmq.stream.Environment
import com.rabbitmq.stream.OffsetSpecification
import com.rabbitmq.stream.ProducerBuilder
import org.springframework.amqp.rabbit.annotation.RabbitListener
import org.springframework.amqp.rabbit.listener.RabbitListenerContainerFactory
import org.springframework.amqp.support.converter.Jackson2JsonMessageConverter
import org.springframework.amqp.support.converter.MessageConverter
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Primary
import org.springframework.rabbit.stream.config.StreamRabbitListenerContainerFactory
import org.springframework.rabbit.stream.listener.ConsumerCustomizer
import org.springframework.rabbit.stream.listener.StreamListenerContainer
import org.springframework.rabbit.stream.producer.RabbitStreamTemplate


/**
 *
 */
@Configuration
class RabbitStreamConfig() {

    @Value("\${spring.rabbitmq.username:guest}")
    private var username: String = "guest";

    @Value("\${account.generator.stream.name:event-streaming-showcase.stream-account-geode-sink-stream}")
    private val streamName: String = "event-streaming-showcase.stream-account-geode-sink-stream"

    @Value("\${spring.rabbitmq.password:guest}")
    private var password:  String = "guest";

    @Value("\${spring.rabbitmq.host:localhost}")
    private var hostname: String = "localhost";

    @Value("\${spring.application.name}")
    private var applicationName: String? = null


    @Bean
    fun rabbitStreamEnvironment(): Environment {

        return Environment.builder()
            .host(hostname)
            .username(username)
            .password(password).build()

    }
    @Bean
    fun streamTemplate(env: Environment): RabbitStreamTemplate {
        val template = RabbitStreamTemplate(env, streamName)
        template.setProducerCustomizer { name: String?, builder: ProducerBuilder ->
            builder.name(
                "test"
            )
        }

        template.setMessageConverter(messageConverter())

        return template
    }

    @Bean
    fun nativeFactory( env : Environment)  : RabbitListenerContainerFactory<StreamListenerContainer>{
        var factory =  StreamRabbitListenerContainerFactory(env);
        factory.setNativeListener(true);


        env.streamCreator().stream("test.stream.queue1").create()

        val cc : ConsumerCustomizer = ConsumerCustomizer{ id: String, builder: ConsumerBuilder ->
            builder.name("myConsumer")
                .offset(OffsetSpecification.first())
                .manualTrackingStrategy();
        }
        factory.setConsumerCustomizer(cc);
        return factory;
    }

    @Primary
    @Bean
    fun messageConverter() : MessageConverter
    {
        return Jackson2JsonMessageConverter();
    }

//    @RabbitListener(queues = ["test.stream.queue1"],containerFactory = "nativeFactory")
//    fun streamListener(input : String) {
//    }
//
//    @RabbitListener(id = "test", queues = ["test.stream.queue2"], containerFactory = "nativeFactory")
//    void nativeMsg(Message in, Context context) {
//        ...
//        context.storeOffset();
//    }

}